### PR TITLE
1.add pull the menu;

### PR DESCRIPTION
**Pull request by Simen-z**
1.add pull the menu.That's what I was talking about before with the drop-down menu, or if you have a better name, use the example:

```
lv_obj_t *pull_win = NULL;
...
        /* create main window*/
        lv_obj_t *main_win = lv_obj_create(lv_disp_get_scr_act(NULL), NULL);
	lv_obj_set_style(main_win , &lv_style_pretty_color);
	lv_obj_set_size(main_win, 416, 416);
        lv_obj_set_event_cb(main_win, event_cb);

        /* create drop-down menu*/
	static lv_style_t pullStyle;
	lv_style_copy(&pullStyle, &lv_style_plain);
	pullStyle.body.main_color = LV_COLOR_BLUE;
	pullStyle.body.grad_color = LV_COLOR_BLUE;
	pullStyle.body.opa = 50;
	pull_win = lv_pull_win_create(lv_disp_get_scr_act(NULL), NULL);
	lv_pull_win_set_size(pull_win , 416, 416);
	lv_obj_set_drag_dir(pull_win , LV_DRAG_DIR_VER);
	lv_pull_win_set_style(pull_win , LV_PULL_WIN_STYLE_BG, &pullStyle);
	
         /* Add a button to the drop-down menu*/
	lv_obj_t * pullLabel;
	lv_obj_t * pullBtn = lv_btn_create(lv_pull_win_get_obj(pull_win), NULL);
	lv_obj_align(pullBtn, NULL, LV_ALIGN_CENTER, 0, -40);
	pullLabel = lv_label_create(pullBtn, NULL);
	lv_label_set_text(pullLabel, "Button");
         /* hide the drop-down menu*/
	lv_pull_win_set_show(pull_win , false, false);
...

static void event_cb(lv_obj_t * obj, lv_event_t event)
{
	lv_point_t curPoint;
	lv_indev_t * pIndev = lv_indev_get_act();
	if (pIndev == NULL)
		return;

	lv_indev_get_point(pIndev, &curPoint);
	switch (event){
	case LV_EVENT_PRESSED:
		if (curPoint.y < 20){
                        /* If y is less than 20, move the drop-down menu to the foreground */
			lv_obj_move_foreground(pull_win);
		}
		break;
	}
}
```

### CODE REVIEW

**Comment by kisvegabor** [comment]
Thank you!

I've tried it and it worked well for me.

However, the `tile` part wasn't clear to me. Can you show an example about how to use them?

File: src/lv_objx/lv_pull_win.h [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+#include "lv_conf.h"
+#else
+#include "../../../lv_conf.h"
+#endif
+
+#if LV_USE_PULL_WIN != 0
+
+#include "../lv_objx/lv_page.h"
+
+	/*********************
```
Comment (embeddedt) [inline src/lv_objx/lv_pull_win.h]: Can you remove the extra indentation here, please?

**Comment by Simen-z** [comment]
@kisvegabor This doesn't have anything to do with tile, it's just that this object I rewrote from the tile object, there might be some inappropriate function and variable names, sorry.Of course, you can also use it in tile, using examples:
```
        lv_obj_t *pull_win = NULL;
        lv_obj_t *objTileview = NULL;
...	
        lv_coord_t hres = 416;
	lv_coord_t vres = 416;

	static const lv_point_t vp[] = {
			{ 0, 0 },//page0
			{ 1, 0 },//page1
			{ 2, 0 },//page2
	};

	objTileview = lv_tileview_create(lv_disp_get_scr_act(NULL), NULL);
	lv_tileview_set_valid_positions(objTileview, vp, 3);
	lv_tileview_set_edge_flash(objTileview, true);
	lv_obj_set_size(objTileview, hres, vres);
	lv_page_set_sb_mode(objTileview, LV_SB_MODE_OFF);
	lv_obj_set_event_cb(objTileview, tileview_event_cb);
	
	/*page0*/
	lv_obj_t *objPage0 = lv_obj_create(objTileview, NULL);
	lv_obj_set_click(objPage0, true);
	lv_obj_set_style(objPage0, &lv_style_pretty_color);
	lv_obj_set_size(objPage0, lv_obj_get_width(objTileview), lv_obj_get_height(objTileview));
	lv_tileview_add_element(objTileview, objPage0);
	lv_obj_set_drag_parent(objPage0, true);
	lv_obj_set_parent_event(objPage0, true);	
	/*page1*/
	lv_obj_t *objPage1 = lv_obj_create(objTileview, objPage0);
	lv_obj_align(objPage1, objPage0, LV_ALIGN_OUT_RIGHT_MID, 0, 0);
	lv_tileview_add_element(objTileview, objPage1);
	/*page2*/
	lv_obj_t *objPage2 = lv_obj_create(objTileview, objPage0);
	lv_obj_align(objPage2, objPage1, LV_ALIGN_OUT_RIGHT_MID, 0, 0);
	lv_tileview_add_element( objPage2);


        /* create drop-down menu*/
	static lv_style_t pullStyle;
	lv_style_copy(&pullStyle, &lv_style_plain);
	pullStyle.body.main_color = LV_COLOR_BLUE;
	pullStyle.body.grad_color = LV_COLOR_BLUE;
	pullStyle.body.opa = 50;
	pull_win = lv_pull_win_create(lv_disp_get_scr_act(NULL), NULL);
	lv_pull_win_set_size(pull_win , hres , vres );
	lv_obj_set_drag_dir(pull_win , LV_DRAG_DIR_VER);
	lv_pull_win_set_style(pull_win , LV_PULL_WIN_STYLE_BG, &pullStyle);
	
         /* Add a button to the drop-down menu*/
	lv_obj_t * pullLabel;
	lv_obj_t * pullBtn = lv_btn_create(lv_pull_win_get_obj(pull_win), NULL);
	lv_obj_align(pullBtn, NULL, LV_ALIGN_CENTER, 0, -40);
	pullLabel = lv_label_create(pullBtn, NULL);
	lv_label_set_text(pullLabel, "Button");
         /* hide the drop-down menu*/
	lv_pull_win_set_show(pull_win , false, false);

        lv_tileview_set_tile_act(objTileview, 0, 0, true);
...

static void tileview_event_cb(lv_obj_t * obj, lv_event_t event)
{
	lv_point_t curPoint;
	lv_indev_t * pIndev = lv_indev_get_act();

	if (pIndev == NULL)
		return;

	lv_indev_get_point(pIndev, &curPoint);
	printf("pressed:%d %d event:%d\n", curPoint.x, curPoint.y, event);

	lv_coord_t xCurPage = 0;;
	lv_coord_t yCurPage = 0;;

	lv_tileview_get_tile_act(objTileview, &xCurPage, &yCurPage);

	switch (event)
	{
	case LV_EVENT_PRESSED:
		if ((curPoint.y < 20) && (xCurPage == 0))
		{
			lv_obj_move_foreground(pull_win);
		}
		break;
	}
}
```

File: src/lv_objx/lv_pull_win.h (hunk 1 above)
Comment (Simen-z) [reply src/lv_objx/lv_pull_win.h]: Oh, I didn't notice. It's already been modified.

**Comment by kisvegabor** [comment]
@LanFengXiMenChuiXue 
 I meant what is the purpose of [these fields](https://github.com/littlevgl/lvgl/pull/1301/files#diff-661c0e982cc4dcc15d721a78d46ac58bR64-R70)?
Or they rae not required?

**Comment by Simen-z** [comment]
these fields?
```
	lv_point_t act_id;
 	uint8_t drag_top_en : 1;
 	uint8_t drag_bottom_en : 1;
 	uint8_t drag_left_en : 1;
 	uint8_t drag_right_en : 1;
 	uint8_t drag_hor : 1;
 	uint8_t drag_ver : 1;
```
They are required. In my example above, I only showed the drop-down menu, but in practice the menu can also be pulled out to the left or right or down.

**Comment by barbiani** [comment]
Is it intended to work like the android?

I've copied the sample and got it to drop when click-hold it.

**Comment by kisvegabor** [comment]
Hi,

Sorry for the late answer.

I was thinking about the widget and it seems to me it's a little bit complex and specific. With the gesture feature, the user can simply create something similar but fully customized.

**Comment by Simen-z** [comment]
Yes, now I also feel it is redundant,you can close this request.

**Comment by kisvegabor** [comment]
Alright, I closed it.
Thank you anyway :slightly_smiling_face:

**Pull request closed without merging**